File: 42_Longest Common Subsequence/Ideal Solutions/longestCommonSubsequence.java
import java.io.*;
import java.lang.*;
import java.util.*;


class Main{
//This function calculates the longest common subsequence between the two strings.
public static int longestCommonSubsequence( char[] arr1, char[] arr2){
    int m = arr1.length;
    int n = arr2.length;
	int dp[][] = new int[m+1][n+1];
	for (int i=0; i<=m; i++){
	for (int j=0; j<=n; j++){
		if (i == 0 || j == 0)
			dp[i][j] = 0;
		else if (arr1[i-1] == arr2[j-1])
			dp[i][j] = dp[i-1][j-1] + 1;
		else
			dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
	    }
	}
  return dp[m][n];
}
//Main function
public static void main(String[] args){
	String s1 = "abcde";
	String s2 = "ace";
	char[] arr1=s1.toCharArray();
	char[] arr2=s2.toCharArray();
    int ans = longestCommonSubsequence(arr1,arr2);
	System.out.println(ans);
  }
}

File: 42_Longest Common Subsequence/Ideal Solutions/longestCommonSubsequence.py

# Finding Longest Subsequence between two sequences
def longestCommonSubsequence(sequenceFirst, sequenceSecond):

    # Declare 2D DP array of size first string and second string
    longestSubsequence = [[0 for i in range(len(sequenceSecond) + 1)] for j in range(len(sequenceFirst) + 1)]

    # Appending Space at beginning , So that loop will begin from 1st index
    sequenceFirst = " " + sequenceFirst
    sequenceSecond = " " + sequenceSecond

    # loop through both sequence
    for row in range(1, len(sequenceFirst)):
        for column in range(1, len(sequenceSecond)):

            # If Char matches then, add 1 to total combinations
            if sequenceFirst[row] == sequenceSecond[column]:
                longestSubsequence[row][column] = 1 + longestSubsequence[row-1][column-1]

            # If not matches then, take max of both remaining subsequence
            else:
                longestSubsequence[row][column] = max(longestSubsequence[row-1][column], longestSubsequence[row][column-1])

    return longestSubsequence[-1][-1]


if __name__ == "__main__":

    # Static Input
    text1 = "abcde"
    text2 = "ace"

    print(longestCommonSubsequence(text1, text2))
